node/src/components/consensus/protocols/highway/round_success_meter.rs

```diff
@@ -0,0 +1,143 @@
+use std::{cmp::max, collections::VecDeque};
+
+use datasize::DataSize;
+
+use crate::{
+    components::consensus::{
+        highway_core::{finality_detector::FinalityDetector, State, Weight},
+        traits::Context,
+    },
+    types::Timestamp,
+};
+
+const NUM_ROUNDS_TO_CONSIDER: usize = 40;
+const MAX_FAILED_ROUNDS: usize = 10;
+const ACCELERATION_PARAMETER: u64 = 1000;
+const THRESHOLD: u64 = 1;
+
+#[derive(DataSize, Debug)]
+pub(crate) struct RoundSuccessMeter<C>
+where
+    C: Context,
+{
+    // store whether a particular round was successful
+    // index 0 is the last handled round, 1 is the second-to-last etc.
+    rounds: VecDeque<bool>,
+    current_round_id: u64,
+    proposals: Vec<C::Hash>,
```

Comment:
Each new collection with unlimited size makes me worry a bit more about potential resource exhaustion attacks. What if an attacker equivocates a gazillion times in a single round and makes `proposals` grow enormously? Even if they don't exhaust our memory they'll at least make us run the finality detector for each entry.
(Not sure how serious this is; doesn't have to be addressed in this PR but let's not forget about it.)

Reply:
We wouldn't run the full finality detector, though, just a single attempt to detect a level-1 summit. But it's still a good point and we should probably take a note of it.

Reply:
>  What if an attacker equivocates a gazillion times in a single round and makes proposals grow enormously

Wouldn't we filter out equivocator messages? And if they are included in the justifications of the honest validators, this will be capped at `n` forks (where `n` is the number of validators) due to the spam protection mechanism?

Reply:
Currently we don't filter them out, do we? And depending on how we defend against fork spam, we might never do so. This week there was a discussion where we agreed it might be easier to just add the spam to the protocol state, and then later garbage collect it.
The endorsement mechanism will allow us to garbage collect all but `n` forks, but that wouldn't shield this collection from spam at all.

It's true that looking for a level-1 summit is _probably_ quite fast. But it would be good to verify that some time. I'd just like to make sure we keep track of all _potential_ security issues. Not something that needs to be addressed in this PR!